### PR TITLE
Set `PortsPerNetworkInterface` when creating a `NATGateway`

### LINKED ISSUE #545: Set `PortsPerNetworkInterface` when creating a `NATGateway`
# Summary

In the current implementation we rely on the default number of ports for a `NetworkInterface` when creating a `NATGateway` in when setting up the infrastructure for a `Shoot`. We need to investigate where it make sense to either:

- Automatically determine a meaningful value here
- Make it configurable via the extension configuration
- Support both of the above

### PR DESCRIPTION
**Pull request by ushabelgur**
# Proposed Changes

- provide the configuration of the portsPerNetworkInterfaces of a NATGateway by introducing a new NATConfig field to the Infrastructure configuration.
- NATConfig will allow configuration of the portsPerNetworkInterfaces of a NATGateway by setting PortsPerNetworkInterface(Same field will be used as MinPortsPerNetworkInterface ) 
- NATConfig will allow to automatically determine a meaningful value if MaxPortsPerNetworkInterface value is also set.

Fixes #545

### CODE REVIEW

File: pkg/controller/infrastructure/actuator_reconcile.go [hunk 1]
```diff
@@ -117,12 +120,50 @@ func (a *actuator) applyNATGateway(ctx context.Context, ironcoreClient client.Cl
...
+				subnet, size := ipv4Net.Mask.Size()
+				amount := big.NewInt(0).Sub(big.NewInt(2).Exp(big.NewInt(2), big.NewInt(int64(size-subnet)), nil), big.NewInt(0))
+				maxPorts := big.NewInt(int64(*maxAvailablePorts))
+				ports := big.NewInt(0).Div(maxPorts, amount)
+
+				if ports.Int64() < int64(*minPortsPerNetworkInterface) {
+					natGateway.Spec.PortsPerNetworkInterface = minPortsPerNetworkInterface
+				} else {
+					natGateway.Spec.PortsPerNetworkInterface = ptr.To(previousPowOf2(int32(ports.Int64())))
+				}
```
Comment (kasabe28) [inline pkg/controller/infrastructure/actuator_reconcile.go:147]: I think we should have check for final `natGateway.Spec.PortsPerNetworkInterface`  value is less than `amount` which is nothing but available ports from `nodeCIDR`

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/controller/infrastructure/actuator_reconcile.go (hunk 1 above)
Comment (ushabelgur) [reply pkg/controller/infrastructure/actuator_reconcile.go:147]: That means,  if the computed ports are smaller than min value configured, do we want to keep min value or we want keep the computed value ? @afritzler/@balpert89   what is your opinion ?

File: pkg/controller/infrastructure/actuator_reconcile.go [hunk 2]
```diff
@@ -117,12 +120,50 @@ func (a *actuator) applyNATGateway(ctx context.Context, ironcoreClient client.Cl
...
 	}
 
+	if natConfig := config.NATConfig; natConfig != nil {
+		minPortsPerNetworkInterface := natConfig.PortsPerNetworkInterface
+		if minPortsPerNetworkInterface != nil {
+			natGateway.Spec.PortsPerNetworkInterface = minPortsPerNetworkInterface
+		}
+		if maxAvailablePorts := natConfig.MaxAvailablePorts; minPortsPerNetworkInterface != nil && maxAvailablePorts != nil {
+			if nodeCIDR := cluster.Shoot.Spec.Networking.Nodes; nodeCIDR != nil {
+				_, ipv4Net, err := net.ParseCIDR(*nodeCIDR)
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/controller/infrastructure/actuator_reconcile.go:130]: We are setting only IPv4 while creating NATGateway, just above this. Do we need explicit check or i can add TODO for now ?
https://github.com/ironcore-dev/gardener-extension-provider-ironcore/blob/1405b6061c809a53db1e84e58a72e2a057467301/pkg/controller/infrastructure/actuator_reconcile.go#L116

File: pkg/controller/infrastructure/actuator_reconcile_test.go [hunk 3]
```diff
@@ -151,8 +159,16 @@ var _ = Describe("Infrastructure Reconcile", func() {
 			},
 			Spec: extensionsv1alpha1.InfrastructureSpec{
 				DefaultSpec: extensionsv1alpha1.DefaultSpec{
-					Type:           ironcore.Type,
-					ProviderConfig: &runtime.RawExtension{Object: &v1alpha1.InfrastructureConfig{}},
+					Type: ironcore.Type,
+					ProviderConfig: &runtime.RawExtension{Object: &v1alpha1.InfrastructureConfig{
+						TypeMeta: metav1.TypeMeta{
+							APIVersion: v1alpha1.SchemeGroupVersion.String(),
```
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/controller/infrastructure/actuator_reconcile_test.go:165]: Do we need to consider a validation if **minPorts** is > **maxPorts**

File: pkg/controller/infrastructure/actuator_reconcile.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/controller/infrastructure/actuator_reconcile.go]: Added IPv4 check before calculation

File: pkg/controller/infrastructure/actuator_reconcile_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (afritzler) [reply pkg/controller/infrastructure/actuator_reconcile_test.go:162]: That would be great! Forgot to mention this in my review.

File: pkg/ironcore/types.go [hunk 4]
```diff
@@ -99,6 +99,8 @@ const (
 	MachineControllerManagerVpaName = "machine-controller-manager-vpa"
 	// MachineControllerManagerMonitoringConfigName is the name of the ConfigMap containing monitoring stack configurations for machine-controller-manager.
 	MachineControllerManagerMonitoringConfigName = "machine-controller-manager-monitoring-config"
+	// MaxAvailableNATPortsPerNetworkInterface defines the maximum number of ports per network interface the NAT gateway should use.
+	MaxAvailableNATPortsPerNetworkInterface = 64512
```
Comment (ushabelgur) [inline pkg/ironcore/types.go:103]: As discussed with @balpert89 max number of ports comes from the linux kernel and the reserved number of ports (1024). so this is always 65536 - 1024 = 64512

File: pkg/controller/infrastructure/actuator_reconcile.go (hunk 1 above)
Comment (ushabelgur) [reply pkg/controller/infrastructure/actuator_reconcile.go]: Added validations to check its always power of 2, greater than zero and always less than max ports

File: pkg/controller/infrastructure/actuator_reconcile_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ushabelgur) [reply pkg/controller/infrastructure/actuator_reconcile_test.go:162]: Added validation

**Review by afritzler — APPROVED** [review]

**Pull request merged**